### ISSUE TITLE
Feature: Basic Lua interop

### ISSUE THREAD

**Issue by ninovanhooff** [samdze/playdate-nim#21]
My Use case: I'm thinking of building a game that is mostly in Nim, but uses comic-style cutscenes. The Panels library is perfect for this: https://cadin.github.io/panels/docs/cutscenes.html

Unfortunately only available in Lua. Basic Lua interop is needed to realise this.
The recommended Panels api might be difficult to call from Nim 
Example: `Panels.startCutscene(comicData, callback)` requires to call a function in a "namespace" (a global lua table, really) and requires two complex parameters.

I propose a more practical approach:
1. In Lua, call the functions as the library offers them, and expose that as a global Lua function without parameters and return values
``` lua

isCutsceneRunning = false

local cutsceneFinishedCallback = function()
  isCutSceneRunning = false
end

function getCutsceneRunning()
  return isCutsceneRunning
end

function startChapter2Cutscene()
  local comicData = [..]
  Panels.startCutscene(comicData, callback)
  isCutsceneRunning = true
end

-- nice to have / stretch goal: support for simple parameter types like int and bool:

function startChapterCutScene(chapterNumber)
  Panels.startCutscene(comicDatas[chapterNumber], callback)
  isCutsceneRunning = true
end
```

2. In Nim, call the lua function to start a cutscene when necessary:
```nim
let (result, errorMessage) = playdate.lua.callFunction("startChapter2Cutscene")

proc update(): int
  var (result, errorMessage) = playdate.lua.callFunction("getCutsceneRunning")
 (result, errorMessage) = playdate.lua.callFunction("updatePanels")
  return 1
```

hmmm. At this point I realise that the playdate api doesn't allow for reading lua globals (`isCutSceneRunning`), or getting a return value from a lua function. In the example above, result is just an int indicating whether an error occurred.

If I'm not missing something, this implies we would also need to implement the playdate.lua.addFunction so that we can

```nim
var isPanelsCutsceneRunning = false

proc panelsCutsceneFinished()
  isPanelsCutsceneRunning = false

proc handler(event: PDSystemEvent, keycode: uint) {.raises: [ValueError].} =

    if event == kEventInitLua:
        let addFuncError = playdate.lua.addFunction(panelsCutsceneFinished, "onPanelsCutsceneFinished")
```

We can then call this from lua:
```lua
function startChapter2Cutscene()
  local comicData = [..]
  Panels.startCutscene(comicData, onPanelsCutsceneFinished)
end
```

So, in summary, I would like to have available:
- playdate.lua.addFunction(someProc). For my usecase no parameters and no return types are needed for someProc)
- playdate.lua.callFunction. For this, no arguments need to be passed to the lua function, although it would nice to have

I would encourage implementing the outerror that the playdate api returns, so that I as a developer can check whether the calls were succesful

**Comment by samdze** [comment]
I will look into this, Lua interop would definitely be useful

**Comment by samdze** [comment]
Update on this:
it's not possible to call Lua function from Nim (or from C) without initializing the Lua environment.
And unfortunately, seems like the Lua environment can be initialized only by not defining the main game loop in Nim (or again, in C).

So, the current situation could be like this:

1. A Lua defined main loop game can call Lua and native C/Nim functions.
C/Nim functions defined in a Lua defined main loop game can call Lua functions.

2. A C/Nim defined main loop game can't call Lua functions because the Lua environment doesn't get initialized.

**Comment by ninovanhooff** [comment]
Good news!

I have already achieved a nim game loop + initialized Lua environment.

It is a matter of setting the C update callback in the InitLua event. You
can ignore the regular init event completely or use it to setup parts of
your nim code.

I can share a proof of concept if you want, but I found it to be a 1 line
code change of handling the InitLua event instead of the Init event.

I was then able to have a running Nim loop with a working print statement
in main.lua

**Comment by samdze** [comment]
Oh, nice.
So yes, just tested, `callFunction` from Nim already works in my local updated branch!

**Comment by samdze** [comment]
I defined `callFunction` as a throwing function, at low level you'd use it like this:
```nim
try:
    # Function with no arguments.
    playdate.lua.callFunction("helloLua")

    # Function with 1 argument.
    playdate.lua.pushString("HELLO FROM NIM")
    playdate.lua.callFunction("funcWithArg", argsCount = 1) # argsCount label is optional
except:
    playdate.system.logToConsole(getCurrentExceptionMsg())
```

**Comment by ninovanhooff** [comment]
Excellent, great progress! Any idea whether addFunction is feasable?

Otherwise, there might be an unholy way to make it work with just callFunction: set a result var to true, then call a Lua function. In the exception handler you can set the result var to false. You can then return normally on the Lua side to return true or intentionally throw an exception to return false

**Comment by samdze** [comment]
`addFunction` is going to work no problem!
Also, directly passing compatible Nim functions as callback arguments to callFunction (`addFunction` not needed) will be supported!

**Comment by samdze** [comment]
Just pushed the updates!
Now you can define a Nim function like this:
```nim
proc nimCallback(state: LuaStatePtr): cint {.cdecl, raises: [].} =
    let argCount = playdate.lua.getArgCount()
    playdate.system.logToConsole(fmt"Nim callback with {argCount} argument(s)")

    try: 
        for i in countup(1, argCount): # Lua indices start from 1...
            let argType = playdate.lua.getArgType(i.cint)

            case argType:
                of kTypeBool:
                    let value = playdate.lua.getArgBool(i)
                    playdate.system.logToConsole(fmt"Argument {i} is a bool: {value}")
                of kTypeFloat:
                    let value = playdate.lua.getArgFloat(i)
                    playdate.system.logToConsole(fmt"Argument {i} is a float: {value}")
                of kTypeInt:
                    let value = playdate.lua.getArgInt(i)
                    playdate.system.logToConsole(fmt"Argument {i} is an int: {value}")
                of kTypeString:
                    let value = playdate.lua.getArgString(i)
                    playdate.system.logToConsole(fmt"Argument {i} is a string: {value}")
                of kTypeNil:
                    let isNil = playdate.lua.argIsNil(i)
                    playdate.system.logToConsole(fmt"Argument {i} is nil: {isNil}")
                else:
                    playdate.system.logToConsole(fmt"Argument {i} is not a recognized type.")
    except:
        playdate.system.logToConsole(getCurrentExceptionMsg())
    return 0
```

And use it as an argument in a function:
```nim
try:
    playdate.lua.pushFunction(nimCallback)
    playdate.lua.callFunction("callNimCallback", 1)
except:
    playdate.system.logToConsole(getCurrentExceptionMsg())
```

`main.lua`:
```lua
function callNimCallback(arg)
    print("Lua calling Nim callback...")
    arg(1, true, "Hey, this is Lua", 39.0, nil)
end
```

**Comment by ninovanhooff** [comment]
Great, I got it to work!

Here is the sample project that first shows a comic in Lua, then starts a (super minimal) snake game: https://github.com/ninovanhooff/Nim-Snake-Playdate/commit/58783c136f49fedeb3392d72de6b607988c34dcf
